ssl/ssl_client_hello_test.cc

```diff
@@ -345,6 +348,175 @@ TEST(SSLClientHelloTest, ClientHelloKnownExtensions) {
+  // Note: This is a bit artificial as the library might add some extensions
+  // by default. We rely on the callback to check the result.
+  SSL_CTX_set_options(client_ctx.get(), SSL_OP_NO_TICKET);
+
+  bool callback_called = false;
+  ExtensionsPresentTestArgs args = {&callback_called,
+                                    false /* expect_session_ticket */};
+  SSL_CTX_set_client_hello_cb(
+      server_ctx.get(), callback_SSL_client_hello_get1_extensions_present_impl,
+      &args);
+
+  UniquePtr<SSL> client, server;
+  ASSERT_TRUE(ConnectClientAndServer(&client, &server, client_ctx.get(),
+                                     server_ctx.get()));
+  EXPECT_TRUE(callback_called);
+}
+
+// Test SSL_client_hello_get_extension_order to verify its behavior with
+// different buffer sizes and to ensure it correctly reports the number of
+// extensions.
+TEST(SSLClientHelloTest, GetExtensionOrder) {
+  UniquePtr<SSL_CTX> client_ctx(SSL_CTX_new(TLS_method()));
+  UniquePtr<SSL_CTX> server_ctx =
+      CreateContextWithTestCertificate(TLS_method());
+  ASSERT_TRUE(client_ctx);
+  ASSERT_TRUE(server_ctx);
+
+  bool callback_called = false;
+  SSL_CTX_set_client_hello_cb(
+      server_ctx.get(),
+      [](SSL *ssl, int *al, void *arg) -> int {
+        bool *called = static_cast<bool *>(arg);
+        *called = true;
+
+        size_t num_extensions = 0;
+        // First, call with a null buffer to get the count of extensions.
+        if (SSL_client_hello_get_extension_order(ssl, nullptr,
+                                                 &num_extensions) != 1) {
+          return SSL_CLIENT_HELLO_ERROR;
+        }
+        EXPECT_GT(num_extensions, 0u);
+
+        // Allocate a buffer of the correct size and get the extensions.
+        uint16_t* exts =
+            (uint16_t *)OPENSSL_zalloc(sizeof(uint16_t) * num_extensions);
+        if (exts == nullptr) {
+          return SSL_CLIENT_HELLO_ERROR;
+        }
+        if (SSL_client_hello_get_extension_order(
+                ssl, exts, &num_extensions) != 1) {
+          OPENSSL_free(exts);
+          return SSL_CLIENT_HELLO_ERROR;
+        }
+
+        unsigned legacy_version = SSL_client_hello_get0_legacy_version(ssl);
+        EXPECT_EQ(legacy_version, (unsigned)TLS1_2_VERSION);
+
+        // Call with a buffer that is too small and confirm it fails.
+        size_t too_small_num_extensions = num_extensions - 1;
+        uint16_t* too_small_exts =
+            (uint16_t *)OPENSSL_zalloc(sizeof(uint16_t) *
+                                       too_small_num_extensions);
+        if (!too_small_exts) {
+          OPENSSL_free(exts);
+          return SSL_CLIENT_HELLO_ERROR;
+        }
+        // Expect failure
+        if (SSL_client_hello_get_extension_order(
+                ssl, too_small_exts, &too_small_num_extensions) != 0) {
+          OPENSSL_free(exts);
+          OPENSSL_free(too_small_exts);
+          return SSL_CLIENT_HELLO_ERROR;
+        }
```

Comment:
Is there any particular reason why we're using if/else statements in the callback instead of the typical `ASSERT/EXPECT` macros? If it's due to the additional memory we have to free, maybe we could consider using smart pointers? 

No strong opinions here, just thought I'd ask since it seemed like there were two ways of testing going on.

Reply:
The ASSERT_xxx macros can't be used in the callback, but it can use the EXPECT_xxx and ADD_FAILURE macros. I added more checks to ensure the test will fail if any errors occur in the callback.